docs/javascript-unit-tests.md

```diff
@@ -0,0 +1,712 @@
+Additional dependencies should be added in the same manner.
+
+```js
+import * as sample from '../../../../cfgov/unprocessed/js/modules/sample.js';
+```
+
+Some test files use `const` declarations to
+`require` scripts instead of `import`,
+because those files were written before `import` was available.
+We prefer to use `import` because it allows for
+[tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/),
+meaning if two modules are importing the same module
+it should only be included in the bundle once,
+whereas with `require` it would be included twice.
+
+A consequence is that variables can't be used in the import path,
+as they prevent Webpack from figuring out which modules are duplicates.
+For example, this snippet shows how a `require` statement
+should be converted to an `import` statement,
+but without including the `BASE_JS_PATH` variable in the file path:
+
+```js
+// works but could duplicate footer-button.js if other files also require it
+const FooterButton = require( BASE_JS_PATH + 'modules/footer-button.js' );
+
+// doesn't work and the build will fail
+import * as FooterButton from BASE_JS_PATH + 'modules/footer-button.js';
+
+// is ugly but it works and supports tree shaking
+import * as FooterButton from '../../../../cfgov/unprocessed/js/modules/footer-button.js';
+```
+
+`import` also provides a benefit in that you can choose specific parts
+of a module so that you only import the dependencies you need.
+For testing purposes, we will typically import the whole module
+to make sure we have full test coverage.
+[Read the `import` reference guide on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import)
+on how to implement `import` for different use cases.
+
+#### The `describe` function
+
+In Jest (whose syntax is based on Jasmine),
+[`describe` blocks](https://jestjs.io/docs/en/api#describename-fn)
+serves as an organizational structures that you can use
+to outline the methods you need in your JS module.
+
+The root `describe` method is where we put
+the name of the JavaScript module we are testing.
+For the sample, the name of our module is `sample`,
+so we set this up on line 4 of `sample-spec.js`:
+
+```js
+describe( 'sample', () => {
+  …
+} );
+```
+
+This module name will appear in your test output in the console
+when the test is running:
+
+![Console output for failing unit test that reads 'sample: should return a string with expected value'](img/js-unit-test-output.png)
+
+More complex tests will have additional `describe` blocks –
+children of the root `describe` block –
+that should correspond with a particular method in the module.
+For example, if we want to add more functionality to our sample JS,
+we could start by writing these tests in `sample-spec.js`:
+
+```js
+describe( 'sample', () => {
+
+    describe( 'gimmeString()', () => {
+
+      it( 'should return a string with expected value', () => {
+        sampleString = 'Shredder';
```

Comment:
`sampleString` needs to be declared `const sampleString = 'Shredder';`.

Reply:
This is an excerpt from the whole file, which includes `let sampleString;` at the top.

As `sampleString` isn't ever reassigned, I guess it might be better to declare it here inside the test with `const`?

Reply:
Either at the top or in the it statement, but it should be declared.